h/services/subscription.py

```diff
@@ -0,0 +1,82 @@
+from sqlalchemy import func
+from sqlalchemy.orm import Session
+from webob.cookies import SignedSerializer
+
+from h.models import Subscriptions
+from h.security import derive_key
+
+
+class InvalidUnsubscribeToken(Exception):
+    ...
+
+
+class SubscriptionService:
+    # Would be nice if this was something to do with the DB model
+    class Type:
+        REPLY = "reply"
+
+        ALL = frozenset([REPLY])
+
+    MISSING_SUBSCRIPTION_ACTIVE = True
+    """Default behavior when we find a missing subscription."""
+
+    def __init__(self, db_session: Session, secret: bytes, salt: bytes):
+        self._db_session = db_session
+        self._token_serializer = SignedSerializer(
+            secret=derive_key(key_material=secret, salt=salt, info=b"h.notification"),
+            salt=None,
+        )
+
+    def get_subscription(self, user_id: str, sub_type: str) -> bool:
+        return self._get_subscription(user_id, sub_type).active
+
+    def set_subscription(self, user_id: str, sub_type: str, active: bool):
+        self._get_subscription(user_id, sub_type).active = active
+
+    def get_subscriptions(self, user_id: str) -> set:
+        return {
+            sub_type
+            for sub_type in self.Type.ALL
+            if self.get_subscription(user_id, sub_type)
+        }
+
+    def set_subscriptions(self, user_id: str, subscriptions: set):
```

Comment:
I find the `subscriptions` name confusing here, maybe `types`?

We could also use `Set[Type]` for the type anno.